Linters e WM (Aula 9) (#10)
guia-de-sobrevivencia-do-programador/Notas-de-Aula: 9 - Potpourri/README.md

# 9 - Potpourri

> Outras ferramentas e tópicos que não tivemos tempo de fazer uma aula própria.

Reviewer: ```suggestion
Nesta aula apresentamos algumas ferramentas que também são importantes no dia a dia do programador, mas que não puderam ser incluídos em uma aula própria durante o curso.
```

## Window manager

### Desktop Environment (DE)

> Uma DE permite que o usuário tenha uma experiência completa para a interface gráfica do usuário, ícones, barras de ferramentas, papéis de paredes e seu próprio gerenciador de janelas.

Reviewer: ```suggestion
Um *desktop environment* ("ambiente de área de trabalho" ou "ambiente de desktop", abreviado aqui como DE) é um conjunto de programas que implementam a metáfora da "área de trabalho" e que compartilham uma mesma interface gráfica de usuário (GUI), incluindo ícones, barras de ferramentas, papéis de paredes e até mesmo seu próprio *gerenciador de janelas*.
```

> Pode ser comparado com o que é visto no Windows a barra de ferramentas, barra de pesquisa, gerenciador de tarefas etc....

Reviewer: O termo *desktop environment* define tanto o estilo das interfaces de usuário que implementam essa metáfora de ambiente de área de trabalho quanto os programas que implementam essa metáfora. Um DE tipicamente define os ícones, janelas barras de ferramentas, diretórios/pastas, papéis de parede e outros *desktop widgets* (programas que se integram à interface, como relógios, calendários, etc.).

> Cada distribuição de Linux pode vir ou não com uma DE, dependo da distribuição ela te dá várias opções de DEs como pode ser visto na imagem abaixo:

Reviewer: Há vários DEs diferentes que podem ser utilizados com o Linux e o usuário pode escolher aquele que melhor atender suas necessidades. Cada distribuição Linux normalmente opta por um DE para ser oferecido por padrão (ou opta por não oferecer um por padrão) e o personaliza para a distribuição.

<p align="center">
    <img src="./assets/Manjaro_exemplos_de.png" alt="Manjaro Exemplos DE" width="400" height="200" />
</p>

> Essa imagem dá a opção para o usuário baixar a distribuição [Manjaro](https://manjaro.org/download/) com 3 opções oficiais de DEs.

Reviewer: ```suggestion
Algumas distribuições permitem que você escolha qual DE você deseja utilizar antes mesmo de você baixar a distribuição. A figura acima mostra as 3 opções de DEs oferecidas oficialmente pela distribuição Linux [Manjaro](https://manjaro.org/download/).
```

> Geralmente uma DE possui outros aplicativos que ajudam a compôr a interface gráfica do usuário.

Reviewer: ```suggestion
```

Reviewer: Isso já foi dito antes, eu acho.

> Alguns DEs famosos são: `KDE`, `GNOME`, `XFCE`, `MATE`, etc...

Reviewer: ```suggestion
Alguns exemplos de DEs famosos: [GNOME](https://www.gnome.org/), [KDE](https://kde.org/pt-br/), [MATE](https://mate-desktop.org/), [XFCE](https://xfce.org/) e [LXDE](http://www.lxde.org/).
```

> O usuário tem a liberdade de adicionar programas de outros DEs por exemplo usando `KDE` eu posso usar programas do `GNOME` (provavelmente o tema dos programas não vai ser o mesmo já que são de DEs diferentes)

Reviewer: ```suggestion
O usuário tem a liberdade de usar programas feito para um DE em outros DEs. Por exemplo, um usuário que esteja usando o `KDE` pode usar programas pensados para o `GNOME` e vice-versa, ainda que provavelmente o programa tenha um aspecto (*look and feel*) diferente do DE utilizado (o tema, ou seja, a escolha de cores, de ícones, etc. do programa será diferente do tema do DE usado).
```

> [Lista](https://wiki.archlinux.org/title/Desktop_environment_%28Portugu%C3%AAs%29#Lista_de_ambientes_de_desktop) de DEs disponíveis.

Reviewer: ```suggestion
O [wiki do Arch Linux](https://wiki.archlinux.org/) (que, diga-se de passagem, possui uma documentação excelente sobre tudo que relacionado à Linux) possui uma extensa [lista de DEs](https://wiki.archlinux.org/title/Desktop_environment_%28Portugu%C3%AAs%29#Lista_de_ambientes_de_desktop) para Linux.
```

### Window Manager (WM)

> Um WM ou gerenciador de janelas é o programa que vai controlar o posicionamento e a aparência das janelas em uma interface gráfica (GUI), por ser mais simples que uma DE, normalmente, um WM acaba utilizando menos memória RAM que um DE.

Reviewer: ```suggestion
Um WM ou gerenciador de janelas é o programa que controla o posicionamento e a aparência das janelas em uma interface gráfica (GUI). Por ser mais simples que uma DE completa, normalmente um WM usa menos memória RAM.
```

> O WM pode ser usado dentro de uma DE ou usada de forma independente na imagem abaixo podemos ver dois WM (sway e i3):

Reviewer: ```suggestion
O WM pode ser usado dentro de uma DE ou usada de forma independente. Na imagem abaixo podemos ver dois WMs: [sway](https://swaywm.org/) e [i3](https://i3wm.org/).
```

<p align="center">
    <img src="./assets/Manjaro_exemplos_wm.png" alt="Manjaro Exemplos WM" width="400" height="200" />
</p>

> Cada WM é feito em uma determinada linguagem, ter o conhecimento prévio dessa linguagem pode ajudar bastante para poder configurar o seu WM da maneira que desejear.

Reviewer: ```suggestion
Cada WM é implementado em uma determinada linguagem. Ter conhecimento prévio dessa linguagem pode ajudar bastante para poder configurar o seu WM da maneira que desejar.
```

Alguns exemplos:

- `awesome` é bom ter um conhecimento em lua.
- `dwm` é bom ter um conhecimento em C.
- `qtile` é bom ter um conhecimento em python.

> - `xmonad` é bom ter um conhecimento em haskell.

Reviewer: ```suggestion
- [awesome](https://awesomewm.org/) pode ser personalizado com código escrito na linguagem de programação Lua;
- [dwm](https://dwm.suckless.org/) requer que seu código-fonte em C seja modificado para ser personalizado;
- [qtile](http://www.qtile.org/) pode ser personalizado com código em Python;
- [`xmonad`](https://xmonad.org/) pode ser personalizado com código em Haskell.
```

#### Tipos

Os WMs podem ser classificados em 3 tipos:

* [Stacking](https://wiki.archlinux.org/title/Window_manager_%28Portugu%C3%AAs%29#Gerenciadores_de_janela_de_empilhamento): É o tipo de organização mais comum para usuários de DEs, Windows e OS X, esse tipo de WM os programas ficam `flutuando` pela área de trabalho.

* [Tiling](https://wiki.archlinux.org/title/Window_manager_%28Portugu%C3%AAs%29#Gerenciadores_de_janela_de_tiling): Esse tipo de WM tem uma organização `automática`. Quando novas janelas são criadas elas automaticamentes são organizadas na área de trabalho de uma forma que as janelas não vão se sobrepor, a organização das janelas vai depender do WM usado ou da configuração do WM.

* [Dynamic](https://wiki.archlinux.org/title/Window_manager_%28Portugu%C3%AAs%29#Gerenciadores_de_janela_din%C3%A2micos): Esse tipo de WM abrange os dois tipos mostrados anteriormente o Stacking e o Tiling, O Dynamic consegue `transicionar` entre os modos de Stacking e Tiling, assim fica à escolha do usuário em qual modo ele quer que as janelas fiquem.

### WM x DE

> Como visto antes uma DE é um WM com outras coisas que integram o desktop para que ele consiga ter bastantes funcionalidades.

Reviewer: ```suggestion
Como visto anteriormente, uma DE é um WM com funcionalidades e programas adicionais que se integram e se complementam para implementar a metáfora de ambiente área de trabalho.
```

> Enquanto o WM é só um gerenciador de tarefas, ou seja ele pode ou não ter coisas como um sistema de menu, notificações, gerenciador de clipboard, gerenciador de internet, gerenciador de login, etc... outros softwares que vão ficar responsáveis por gerenciar essas coisas. Mas como o WM vem sem nada o usuário pode escolher o que ele quer instalar e vai gastar menos memória RAM no seu computador.

Reviewer: ```suggestion
Normalmente um WM atua exclusivamente como um gerenciador de janelas, ou seja, tipicamente um WM não implementa coisas como um sistema de menu, notificações, gerenciador de *clipboard*, gerenciador de redes, gerenciador de *login*, etc. Outros softwares (opcionais) são responsáveis por gerenciar essas coisas. Como um WM oferece menos recursos, optar por usar apenas um WM pode economizar bastante memória RAM do seu computador.
```

> Para quem usa bastante o teclado vai ser uma boa opção já que quase nunca você vai precisar tirar suas mãos do teclado.

Reviewer: ```suggestion
Muitos WMs permitem que o usuário realize todas as operações de gerenciamento de janelas usando apenas o teclado. Usuários proficientes conseguem trabalhar em tais WMs sem tirar as mãos do teclado, o que acelera bastante o trabalho.
```

> ### Complemento

Reviewer: ```suggestion
### Complementos
```

> Como os WM não tem todos os programas necessários para ser tão funcional como uma DE alguns programas abaixo podem ser úteis na configuração de um WM:

Reviewer: ```suggestion
Dependendo das necessidades do usuário, pode ser preciso complementar a WM com utilitários que forneçam algumas das funcionalidades adicionais encontradas em DEs:
```

- Inicializador de programas: `dmenu` ou `rofi`
- Panel: `polybar`

> - Notificações: `dunst`

Reviewer: ```suggestion
- Inicializador de programas: [dmenu](https://tools.suckless.org/dmenu/) ou [rofi](https://github.com/davatorium/rofi)
- Painel: [polybar](https://github.com/polybar/polybar)
- Notificações: [dunst](https://github.com/dunst-project/dunst)
```

## Linters

### O que é um linter?

Linter é uma ferramenta para ajudar no desenvolvimento do seu código. O Linter analisa o código especificado procurando por problemas, podem ser eles problemas técnicos ou problemas de qualidade de código, sendo assim capazes de diagnosticar e arrumar problemas que podem passar despercebidos pelo programador.

O resultado será um código mais legível, consistente e mais fácil de manter, dessa forma também estarão contribuindo para ajudar programadores mais inexperientes à aprender sobre qualidade de código.

Além disso alguns Linters também conseguem manter o código mais seguro e perfomático, mas isso não é uma regra.

### Análise Estática

A análise estática é feito pelos linters para analisar o código em questão sem executar ele, checando por problemas, bugs, vazamento de memória e outros problemas.

### Erros de Sintaxe

Uma das coisas mais básicas que o Linter pode oferecer são verificar erros de sintaxe.

> Erros de sintaxe são detectadas pelos compiladores em linguagens `compiladas` como c, c++, java, etc... Mas para linguagens `interpretadas` os erros apenas vão ser encontrados quando o código for executado, é aí que o Linter entra!

Reviewer: ```suggestion
Erros de sintaxe são detectadas pelos compiladores em linguagens `compiladas` como C, C++, Java, etc. Mas em linguagens *interpretadas*, os erros serão encontrados apenas quando o código for executado, e é aí que o uso de um Linter se torna importante.
```

Um exemplo pode ser erros do `CSS` que vão apenas ser percebidos quando o site for acessado.

### Convenções de código

Convenções de código são de extrema importância quando se quer criar um código para ser mantido ou um cógido que vai ser usado por muitas pessoas.

Ter uma convenção de código único facilita a leitura de um código, ao invés de ter estilos de código de vários desenvolvedores diferentes que iria dificultar a legibilidade do código.

> 

Reviewer: ```suggestion

Uma lista de convenções de códigos para diferentes linguagens de programação e para projetos de desenvolvimento de software pode ser vista nesta página da Wikipédia: https://en.wikipedia.org/wiki/Coding_conventions#List_of_coding_standards .
```

Muitos linters também incluem verificação de performace, eles podem adicionar diferentes tipos de performace e para novos programadores pode ajudar bastante no desenvolvimento e no aprendizado.

### Segurança

A segurança é um dos maiores problemas no desenvolvimento de um código ou o maior problema que os devem se preocupar.

Não se aplica para todos os linters, mas existem linters específicos que podem oferecer verificações de segurança durante o desenvolvimento de um aplicativo.

### Exemplos

Para quem quiser dar uma olhada nos linters existentes e quais linguagens de programação eles conseguem suportar existe um [repositorio](https://megalinter.github.io/5.7.0/all_linters/) com uma variedade de linters especificando sua licença, linguagem, versão, repositório, etc... Existem outros repositórios com indicações de linters como [esse](https://github.com/caramelomartins/awesome-linters).

Vai depender do programador pesquisar e escolher qual linter é melhor para seu uso no dia-a-dia.

## Suite de redes Linux

Antes de falar sobre a suite de ferramentas para redes em Linux, uma breve aula sobre redes:

> A internet é a rede mais famosa que conhecemos, mas é somente uma de varias redes que regem a conexão entre computadores.

Reviewer: ```suggestion
A Internet é a rede mais famosa que conhecemos, mas é somente uma de várias redes que regem a conexão entre computadores.
```

> Cada rede tem seus próprios endereços IPs, e pode ou não estar conectada a internet ou a outra rede. Esses endereços são o que identificam uma máquina ou um nó na rede. Cada máquina em uma rede tem pelo menos um endereço, e cada endereço é único dentro dessa rede.

Reviewer: ```suggestion
Cada rede tem seus próprios endereços IPs, e pode ou não estar conectada à Internet ou à outra rede. Esses endereços são o que identificam uma máquina ou um nó na rede. Cada máquina em uma rede tem pelo menos um endereço, e cada endereço é único dentro dessa rede.
```

Por que eu digo "pelo menos"? Porque uma mesma máquina pode ter mais de um endereço vinculado a ela, pois o que retém controle desse endereço não é a máquina ou o sistema operacional, mas sim os dispositivos de rede. Se você tiver um adaptador *WiFi* e um cabo conectados ao seu computador ao mesmo tempo, cada uma dessas entradas (ou *NICs* de *Network Interface Cards*) terão seu próprio endereço de IP distinto.

Uma máquina também pode estar conectada a duas redes ao mesmo tempo. Essa máquina é chamada de *gateway*, pois vira o portão, ou *gateway*, entre uma rede e outra (mas pode ter somente um *gateway* por rede, e esse normalmente é o seu roteador, que conecta sua rede local à internet). Se uma máquina em uma dessas redes quiser se conectar a uma máquina da outra rede, a conexão entre essas duas máquinas vão passar por essa máquina configurada como *gateway*.

> Existem vários grupos de endereços que são reservados. Alguns desses grupos são reservados para redes locais, detalhados no RFC 1918. Essas são:

Reviewer: ```suggestion
Existem vários grupos de endereços que são reservados. Alguns desses grupos são reservados para redes locais, detalhados no [RFC 1918](https://datatracker.ietf.org/doc/html/rfc1918). Essas são:
```
